library/bigip_software.py

```diff
@@ -212,673 +290,1060 @@
-                    imageIDs=[image]
-                )
-                result['software'].append(info[0])
-
-            hotfixes = self.api.System.SoftwareManagement.get_software_hotfix_list()
-            for hotfix in hotfixes:
-                info = self.api.System.SoftwareManagement.get_software_hotfix(
-                    imageIDs=[hotfix]
-                )
-                result['hotfix'].append(info[0])
-        except bigsuds.ServerError:
-            pass
-
+            active_volume = str(active).split('.')
+            active_volume[-1] = str(int(active_volume[-1]) + 1)
+            target_volume_name = '.'.join(active_volume)
+        except Exception:
+            target_volume_name = str(active) + '.1'
+
+        self._values['create_volume'] = True
+        return target_volume_name
+
+    def to_return(self):
+        result = {}
+        for returnable in self.returnables:
+            result[returnable] = getattr(self, returnable)
+        result = self._filter_params(result)
         return result
 
-    def get_active_volume(self):
-        softwares = self.api.System.SoftwareManagement.get_all_software_status()
-        for software in softwares:
-            if software['active']:
-                return software['installation_id']['install_volume']
-        return None
-
-    def get_inactive_volumes(self):
-        status = self.api.System.SoftwareManagement.get_all_software_status()
-        volumes = [x['installation_id']['install_volume'] for x in status if not x['active']]
-        volumes.sort()
-        return volumes
+    def api_params(self):
+        result = {}
+        for api_attribute in self.api_attributes:
+            if self.api_map is not None and api_attribute in self.api_map:
+                result[api_attribute] = getattr(self,
+                                                self.api_map[api_attribute])
+            else:
+                result[api_attribute] = getattr(self, api_attribute)
+        result = self._filter_params(result)
+        return result
 
-    def get_next_available_volume(self, delete_target=False):
-        target_volume = None
-        active_volume = self.get_active_volume()
 
-        for volume in self.get_inactive_volumes():
-            if volume == active_volume:
-                continue
-            else:
-                target_volume = volume
-                break
+class ModuleManager(object):
+    def __init__(self, client):
+        self.client = client
+        self.want = Parameters()
+        self.want.client = self.client
+        self.want.update(self.client.module.params)
 
-        if target_volume is None:
-            try:
-                _active_volume = active_volume.split('.')
-                _active_volume[-1] = str(int(_active_volume[-1]) + 1)
-                target_volume = '.'.join(_active_volume)
-            except Exception:
-                target_volume = active_volume + '.1'
+    def exec_module(self):
```

Comment:
they are mutually exclusive, this will not work, that is how it should work, you cannot combine remote source and local source sorry, this adds too much complexity and a whole set of new problems and i don't intend to have it like that